sdk/core/core-https/src/nodeHttpsClient.ts

```diff
@@ -0,0 +1,293 @@
+    super();
+    this.progressCallback = progressCallback;
+  }
+}
+
+/**
+ * A HttpsClient implementation that uses Node's "https" module to send HTTPS requests.
+ */
+export class NodeHttpsClient implements HttpsClient {
+  /**
+   * Makes a request over an underlying transport layer and returns the response.
+   * @param request The request to be made.
+   */
+  public async sendRequest(request: PipelineRequest): Promise<PipelineResponse> {
+    const abortController = new AbortController();
+    let abortListener: ((event: any) => void) | undefined;
+    if (request.abortSignal) {
+      if (request.abortSignal.aborted) {
+        throw new AbortError("The operation was aborted.");
+      }
+
+      abortListener = (event: Event) => {
+        if (event.type === "abort") {
+          abortController.abort();
+        }
+      };
+      request.abortSignal.addEventListener("abort", abortListener);
+    }
+
+    if (request.timeout > 0) {
+      setTimeout(() => {
+        abortController.abort();
+      }, request.timeout);
+    }
+
+    if (request.formData) {
+      prepareFormData(request.formData, request);
+    }
+
+    if (!request.skipDecompressResponse) {
+      request.headers.set("Accept-Encoding", "gzip,deflate");
+    }
+
+    let body = request.body;
+
+    if (body && request.onUploadProgress) {
+      const onUploadProgress = request.onUploadProgress;
+      const uploadReportStream = new ReportTransform(onUploadProgress);
+      if (isReadableStream(body)) {
+        body.pipe(uploadReportStream);
+      } else {
+        uploadReportStream.end(body);
+      }
+
+      body = uploadReportStream;
+    }
+
+    if (body) {
+      if (isReadableStream(body)) {
+        request.headers.set("Transfer-Encoding", "chunked");
+      } else if (typeof body === "string") {
+        request.headers.set("Content-Length", body.length);
+      }
+    }
+
+    try {
+      const result = await new Promise<PipelineResponse>((resolve, reject) => {
+        const options = getRequestOptions(request);
+        const req = https.request(options, async (res) => {
+          const headers = getResponseHeaders(res);
+
+          const status = res.statusCode ?? 0;
+          const response: PipelineResponse = {
+            status,
+            headers,
+            request
+          };
+
+          let responseStream = getResponseStream(res, headers, request.skipDecompressResponse);
+
+          const onDownloadProgress = request.onDownloadProgress;
+          if (onDownloadProgress) {
+            const downloadReportStream = new ReportTransform(onDownloadProgress);
+            responseStream.pipe(downloadReportStream);
+            responseStream = downloadReportStream;
+          }
+
+          if (request.streamResponseBody) {
+            response.readableStreamBody = responseStream;
+          } else {
+            response.bodyAsText = await streamToText(responseStream);
```

Comment:
What are your thoughts on making the `streamToText` part of a deserialization policy, so instead we always set the `readableStreamBody` and have access to that?

I feel like this makes it easier for our deserializers to add more validation/transforms if needed.
For example, we could easily add a policy that checks that the number of bytes in the response body matches the content-length. Or have a policy that gunzips the response body as it is streamed from the service. We could even allow an option in the future to give customers direct access to the response in case they wanted to bypass all our magic altogether!

Reply:
I would love to go down this route, but how do we handle the browser case where the request is *always* unzipped before it's given to you, and you can't participate in the stream/deserialization process?

The reason this layer is so thick is because we don't want to expose transport differences between node and the browser to policies in the pipeline.

Reply:
We already do expose transport-level differences though, in that in node.js, you can get access to a readableStream, and in the browser, you get can get access to a Blob.

We'd have to have some policies that only applied for node.js, and some that only applied for browsers. In practice I'd imagine that would mean that we'd create a policy for both runtimes, with one being a no-op if nothing had to be done for that runtime.

Reply:
Hmm I'd like to think about this some more, but I'm definitely open to refactoring the clients into policies